### PR TITLE
Refactor registry system: no direct dependencies; expose standard hash.Hash; be a data carrier.

### PR DESCRIPTION
**Pull request by warpfork**
This diff revisits the registry system to use more standard interfaces for hashers, and puts it to work to reduce the transitive dependencies of go-multihash.

---

### Dependencies:

Previously, depending on the go-mulithash package would create direct
dependencies to several other modules for their various hash function
implementations.  This meant that instead of go-multihash being a
lightweight, easy-to-accept dependency itself, it became something
which would noticably increase the size of your go.mod file,
your package graph, your download sizes during development, and most
concerningly, your compile output size in final products.

Now, there is a registry system (see the `Register` function), and
the main go-multihash package *only* populates the registry with hashes
that are available from the golang standard library by default.
This means you gain no transitive dependencies on other libraries
by importing the go-multihash package, and your binaries will not
be bloated by hashers you don't use.  (Your go.mod file may still
show more repos; but they don't end up in your builds unless you
actually refer to them).

There are now several new packages in `go-multihash/register/*`.
These can be imported to register the hashes in those packages.
If you want all the hashes that were previously available, just make
sure to import "go-mulithash/register/all" somewhere in your program.
(You can register hashes, too, without making PRs to this library;
these packages are just here for convenience and easy use.)

**This is a breaking change** if you used hashes not found in the
golang stdlib, such as blake2 and sha3.  However, to update,
all you need to do is ensure the relevant `go-multihash/register/*`
package is imported anywhere in your program -- an easy change.
A `go-multihash/register/all` package can be imported to get a hasher
registered for all of the same multihash codes as before (but will
correspondingly add the dependency weight back too, of course).

---

### Standard hash.Hash:

Previously, go-multihash had its own definition of a `HashFunc`
interface, and only exposed hashing through the `multihash.Sum` method.
The problem with this was these definitions did not support streaming:
one had to have an entire chunk of memory loaded at once,
in a single contiguous byte slice, in order to hash it.

(A second, admittedly much more minor, problem with this was that one
often had to write glue code to turn a `hash.Hash` into a
`multihash.HashFunc`, and since most of golang uses the standard lib
`hash.Hash` definition already, this was generally avoidable friction.)

Now, the Register function operates in terms of standard `hash.Hash`,
and there is a `GetHasher` function which can get you a `hash.Hash`.
(Okay, to be more precise, these functions take and return a factory
function for a `hash.Hash`.  You get the idea.)

Since the standard `hash.Hash` interface can operate streamingly,
now it's easy to use go-multihash in a streaming way.

The `multihash.Sum` method works the same as always.

---

### Be a data carrier:

Previously, go-multihash contained checks that any multihash indicator
codes being handled were required to have a hash function registered
for them.  This made it very difficult to use go-multihash in a
"forward compatible" way (and it also made a lot of practical bumps
for this dependency-extraction refactor).

Now, go-multihash is willing to carry data, even if it doesn't know
what kind of hash function would be associated with an indicator code.
(Methods that you'd expect to parse things do still parse the varints,
making sure they're sanely formatted.  They just don't inspect and
whitelist the actual integers anymore.)

I removed the `ValidCode` predicate entirely.  It doesn't seem to serve
any good purpose anymore.

---

### Other:

I have not touched the `Codes` and `Names` maps in this diff.
I think we should probably review (and probably remove) these, and
instead direct people to use the go-multicodec package instead,
which has the two advantages of decoupling registration of an
implementation versus simply having a description, and also being
automatically generated from the multiformats table.
However, I wanted to check on feelings about this before doing the work
(especially because they're somewhat entangled with a bunch of the
tests in this package, making their removal somewhat nontrivial).

Most of the test files are now `package multihash_test`.
This makes for some colorful diffs, but is not otherwise interesting.
The reason for this is because the dependency separation process now
requires the tests to import those `register/*` packages, and to
avoid a cycle, that means, well, `package mulithash_test`.

I think there's probably more work to be done in making this library
really shine.  For example, in reviewing the `Encode` function,
I see some allocations that look very likely to be avoidable... if the
function was redesigned to be more aware of how it's likely to be used.
However, I took no action on this, in part because this diff is big
enough already, and in part because I think it might be reasonable to
re-examine the relationship of this code to go-cid at the same time.

I dropped `TestSmallerLengthHashID`.  It appeared to be testing an API
that wasn't actually exported... and the nearest API that *is* exported
(`Sum`) has a general contract of truncating a hash upon short length,
so it was overall unclear what this test should be checking.
Review might be needed on this.

The situation for murmur3 is still in need of resolution.
It's commented out entirely for now.  Questions are noted in the diff.

There's a 'register/miniosha256' package which sets the sha2-256
implementation to a non-stdlib one.  If you don't import this package,
you still get a sha2-256; it's just the stdlib one.
I did not include this in the 'register/all' group.
(Maybe it's faster; maybe it's not; but it's definitely not required,
and I'm getting some reports it also shows weird on profiles, so I tend
to think maybe one should really have to explicitly ask for this one.)

There are a couple other review questions marked with "// REVIEW" in the diff.

### CODE REVIEW

**Comment by warpfork** [comment]
(The test that's currently failing is for murmur3.  Input needed on what we're going to do with that thing.)

File: errata.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	"bytes"
+	"crypto/sha256"
+	"hash"
+)
+
+type identityMultihash struct {
+	bytes.Buffer
+}
+
+func (identityMultihash) BlockSize() int {
```
Comment (Stebalien) [inline errata.go:13]: bytes.Buffer is large, take this by pointer.

File: errata.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+func (x identityMultihash) Size() int {
+	return x.Len()
+}
+
+func (x identityMultihash) Sum(digest []byte) []byte {
+	return x.Bytes()
+}
+
+type doubleSha256 struct {
+	main hash.Hash
```
Comment (Stebalien) [inline errata.go:26]:
```suggestion
	hash.Hash
```

File: errata.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+func (x doubleSha256) Reset() {
+	x.main.Reset()
+}
+
+func (x doubleSha256) Sum(digest []byte) []byte {
+	intermediate := [sha256.Size]byte{}
+	x.main.Sum(intermediate[0:0])
+	h2 := sha256.New()
+	h2.Write(intermediate[:])
```
Comment (Stebalien) [inline errata.go:49]: 1. If we're going to allocate, why not just `sha256.Sum256(intermediate)`?
2. Otherwise, we can probably just re-use the digest instead of creating a new intermediate.

File: multihash.go [hunk 4]
```diff
@@ -267,9 +263,8 @@ func Decode(buf []byte) (*DecodedMultihash, error) {
 // Encode a hash digest along with the specified function code.
 // Note: the length is derived from the length of the digest itself.
 func Encode(buf []byte, code uint64) ([]byte, error) {
-	if !ValidCode(code) {
-		return nil, ErrUnknownCode
-	}
+	// REVIEW: if we remove the strict ValidCode check, this can no longer error.  Change signiture?
```
Comment (Stebalien) [inline multihash.go:266]: I'd leave it. It's slightly annoying but it means we have the error response if we need it and changing this would be more annoying.

File: multihash.go [hunk 5]
```diff
@@ -267,9 +263,8 @@ func Decode(buf []byte) (*DecodedMultihash, error) {
...
+	// REVIEW: this function always causes heap allocs... but when used, this value is almost always going to be appended to another buffer (either as part of CID creation, or etc) -- should this whole function be rethought and alternatives offered?
```
Comment (Stebalien) [inline multihash.go:267]: Later.

File: multihash.go [hunk 6]
```diff
@@ -285,12 +280,6 @@ func EncodeName(buf []byte, name string) ([]byte, error) {
 	return Encode(buf, Names[name])
 }
 
-// ValidCode checks whether a multihash code is valid.
-func ValidCode(code uint64) bool {
```
Comment (Stebalien) [inline multihash.go:289]: known code?

File: register/blake2/multihash_blake2.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+		case 32:
+			multihash.Register(c, blake2b.New256)
+			continue
+		case 64:
+			multihash.Register(c, blake2b.New512)
+			continue
+		}
+
+		// Ok, allocate away.
+		//  (The config object here being a pointer is unfortunate.)
```
Comment (Stebalien) [inline register/blake2/multihash_blake2.go:52]: Can't we allocate it once per function definition then just capture it?

File: register/blake2/multihash_blake2.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+	"golang.org/x/crypto/blake2s"
+
+	"github.com/multiformats/go-multihash"
+)
+
+const (
+	BLAKE2B_MIN = 0xb201
+	BLAKE2B_MAX = 0xb240
+	BLAKE2S_MIN = 0xb241
+	BLAKE2S_MAX = 0xb260
```
Comment (Stebalien) [inline register/blake2/multihash_blake2.go:28]: These should all go in a shared package somewhere.

File: register/miniosha256/multihash_miniosha256.go [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+package miniosha256
+
+import (
+	"github.com/minio/sha256-simd"
+
+	"github.com/multiformats/go-multihash"
+)
+
+func init() {
+	multihash.Register(0x12, sha256.New)
```
Comment (Stebalien) [inline register/miniosha256/multihash_miniosha256.go:22]: (constants...)

File: register/murmur3/multihash_murmur3.go [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+// 	"hash"
+//
+// 	"github.com/gxed/hashland/murmur3"
+//
+// 	"github.com/multiformats/go-multihash"
+// )
+
+func init() {
+	// REVIEW: what go-multihash has done historically is New32, but this doesn't match what the multihash table says, which is 128!  Resolution needed.
+	// REVIEW: I have also heard that something in ipfs unixfsv1 uses a murmur hash, but that is yet different than this.  Resolution needed.
```
Comment (Stebalien) [inline register/murmur3/multihash_murmur3.go:24]: Sharded directories do, but I don't think they use go-multihash. However, they _do_ use the codec.

File: register/sha3/multihash_sha3.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+import (
+	"hash"
+
+	"golang.org/x/crypto/sha3"
+
+	"github.com/multiformats/go-multihash"
+)
+
+func init() {
+	multihash.Register(0x14, sha3.New512)
```
Comment (Stebalien) [inline register/sha3/multihash_sha3.go:25]: We really can't use constants this way.

File: register/sha3/multihash_sha3.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+func (shakeNormalizer) BlockSize() int {
+	return 32 // Shake doesn't have a prefered block size, apparently.  An arbitrary but unsurprising and positive nonzero number has been chosen to minimize the odds of fascinating bugs.
+}
+
+func (x shakeNormalizer) Size() int {
+	return x.size
+}
+
+func (x shakeNormalizer) Sum(digest []byte) []byte {
+	if len(digest) != x.size {
```
Comment (Stebalien) [inline register/sha3/multihash_sha3.go:55]: if `len(digest) < x.size`. Otherwise, just truncate it.

File: sum.go [hunk 13]
```diff
@@ -1,241 +1,49 @@
...
+	hasher, err := GetHasher(code)
+	if err != nil {
+		return nil, err
 	}
 
+	// Feed data in.
+	hasher.Write(data)
+
+	// Compute hash.
+	//  Use a fixed size array here: should keep things on the stack.
```
Comment (Stebalien) [inline sum.go:27]: It won't due to interface indirections (go's escape analysis won't be able to deal with this and will put everything on the heap) and if we don't size this appropriately, CID allocations will 2x.

Let's allocate what we need (use the size hints).

**Review by Stebalien — COMMENTED** [review]
Mostly LGTM but we need to do something about those constants everywhere.

File: register/sha3/multihash_sha3.go (hunk 11 above)
Comment (aschmahmann) [reply register/sha3/multihash_sha3.go:25]: Use go-multicodec seem like a plan?

File: register/murmur3/multihash_murmur3.go (hunk 10 above)
Comment (aschmahmann) [reply register/murmur3/multihash_murmur3.go:24]: See https://github.com/multiformats/go-multihash/issues/135#issuecomment-791178958

I'm pretty sure that we should just remove murmur3 from this library until we can figure out what the correct implementation of the relevant code number is/should be.

**Review by aschmahmann — CHANGES_REQUESTED** [review]
Is it easy enough to fix the binary in the `multihash` folder? Seems like we'd just need it to import all the registry entries here (including miniosha256).

File: errata.go (hunk 1 above)
Comment (warpfork) [reply errata.go:13]: Fixed.

File: errata.go (hunk 2 above)
Comment (warpfork) [reply errata.go:26]: Fixed.

File: errata.go (hunk 3 above)
Comment (warpfork) [reply errata.go:49]: Hm.  Looked at (1), but I think it's slightly, minutely worse to do that because then we can't hand down our `digest` byte slice parameter we were given for output space (and using that slice instead of introducing `intermediate` would be cost-neutral, but isn't a good option either because it would be a break of what people expect Sum to do -- if Sum mutates its parameter it's supposed to be to set it to the result, and I'd be nervous about doing anything else).

I thought about (2), too, but it would make our Sum method side-effecting (ex: two consecutive calls would yield different results) which would also be a break of what Sum is expected to do.

File: register/murmur3/multihash_murmur3.go (hunk 10 above)
Comment (warpfork) [reply register/murmur3/multihash_murmur3.go:24]: Adin's made some more comments about this in https://github.com/multiformats/go-multihash/issues/135 as well.

My (weakly held) position is that we should just flat out remove this.  It sounds like anything currently attempting to use it will get answers that are "wrong" already, so we dearly hope nothing is (or we have data inertia problems coming, which are much _bigger_ problems than a library API break) -- at such a point, I have a hard time seeing any good that can come of its inclusion.

If we do want to keep it, I won't object, but I somebody needs to tell me exactly what code goes here.  I will probably "popcorn" this PR to them to add whatever that diff is.

File: register/blake2/multihash_blake2.go (hunk 8 above)
Comment (warpfork) [reply register/blake2/multihash_blake2.go:28]: I dunno how I feel about these.  I kind of just cargo-culted them over in the first pass.  Looking again now... they seem like an implementation detail that might not be sensible to export at all?  I can't imagine anyone but us looping over these ranges.  (And we don't even use the 2B ranges, at that.)

File: register/blake2/multihash_blake2.go (hunk 7 above)
Comment (warpfork) [reply register/blake2/multihash_blake2.go:52]: Ooh good point.  Improved.

File: register/sha3/multihash_sha3.go (hunk 12 above)
Comment (warpfork) [reply register/sha3/multihash_sha3.go:55]: Fixed.

File: registry.go [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+// The indicator code should be per the numbers reserved and described in
+// https://github.com/multiformats/multicodec/blob/master/table.csv .
+//
+// If Register is called with the same indicator code more than once, the last call wins.
+// In practice, this means that if an application has a strong opinion about what implementation to use for a certain hash
+// (e.g., perhaps they want to override the sha256 implementation to use a special hand-rolled assembly variant
+// rather than the stdlib one which is registered by default),
+// then this can be done by making a Register call with that effect at init time in the application's main package.
+// This should have the desired effect because the root of the import tree has its init time effect last.
+func Register(indicator uint64, hasherFactory func() hash.Hash) {
```
[the remark this answers is not in the document]
Comment (warpfork) [reply registry.go:35]: If we had more arguments that could actually error, sure.  But right now I can't think of any such things.

File: registry.go [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+// The registry automatically contains those hashers which are available in the golang standard libraries
+// (which includes md5, sha1, sha256, sha384, sha512, and the "identity" mulithash, among others).
+// Other hash implementations can be made available by using the Register function.
+// The 'go-mulithash/register/*' packages can also be imported to gain more common hash functions.
+//
+// If an error is returned, it will be ErrSumNotSupported.
+func GetHasher(indicator uint64) (hash.Hash, error) {
+	factory, exists := registry[indicator]
+	if !exists {
+		return nil, ErrSumNotSupported // REVIEW: it's unfortunate that this error doesn't say what code was missing.  Also "NotSupported" is a bit of a misnomer now.
```
[the remark this answers is not in the document]
Comment (warpfork) [reply registry.go:54]: Yeah, I think we should add an error type here, and drop `ErrSumNotSupported`.

The `Sum` function actually previously returned its own fresh unique `fmt.Errorf` that contained the number intead of `ErrSumNotSupported`, which seems like pretty strong evidence that there's no significant legacy/compat reason for keeping `ErrSumNotSupported`.

@Stebalien I'm going to call on you to be the final arbiter of this choice.

File: sum.go [hunk 16]
```diff
@@ -1,241 +1,49 @@
...
+	var space [64]byte
+	sum := hasher.Sum(space[0:0])
+
+	// Deal with any truncation.
 	if length < 0 {
 		var ok bool
 		length, ok = DefaultLengths[code]
```
[the remark this answers is not in the document]
Comment (warpfork) [reply sum.go:34]: Yeah, I just didn't touch this yet.  Input wanted.

Option 3 seems bad to me, Options 1 / 2 / 4 all seem potentially viable.

File: sum_test.go [hunk 17]
```diff
@@ -122,75 +123,22 @@ func BenchmarkBlake2B(b *testing.B) {
 	}
 }
 
-// Test that the identity hash function checks
-// its `length` arguemnt for values that are
-// different to its `data` length.
-func TestSmallerLengthHashID(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (warpfork) [reply sum_test.go:128]: I made a comment at the top about this.

It appeared to be testing an API that wasn't actually exported... and the nearest API that is exported (Sum) has a general contract of truncating a hash upon short length, so it was overall unclear what this test should be checking.

So, I'm game to put this, or something like it, back in.  But it's just not clear to me what it's supposed to be doing -- if I port it in the most literal possible way (e.g. to call `Sum`), it fails, but "fixing" it would change the general contract of `Sum` -- so I'd want someone to chime in on what the expected behavior should actually be.

File: register/blake2/multihash_blake2.go [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+)
+
+func init() {
+	// BLAKE2S
+	// This package only enables support for 32byte (256 bit) blake2s.
+	multihash.Register(BLAKE2S_MIN+31, func() hash.Hash { h, _ := blake2s.New256(nil); return h })
```
[the remark this answers is not in the document]
Comment (warpfork) [reply register/blake2/multihash_blake2.go:34]: They still are.  It's in another init function in the main package.

(That said, this also touches upon the [approach to constants issue](https://github.com/multiformats/go-multihash/issues/137).)

File: errata.go [hunk 19]
```diff
@@ -0,0 +1,51 @@
+package multihash
+
```
[the remark this answers is not in the document]
Comment (warpfork) [reply errata.go:2]: Yep.

Other filename suggestions welcome, if thou doth opine muchly.

File: sum.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (warpfork) [reply sum.go:34]: Maaaybe related to https://github.com/multiformats/go-multihash/issues/137 (if we choose Option 4).

File: registry.go [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+// registry is a simple map which maps a multihash indicator number
+// to a standard golang Hash interface.
+//
+// Multihash indicator numbers are reserved and described in
+// https://github.com/multiformats/multicodec/blob/master/table.csv .
+// The keys used in this map must match those reservations.
+//
+// Hashers which are available in the golang stdlib will be registered automatically.
+// Others can be added using the Register function.
+var registry = make(map[uint64]func() hash.Hash)
```
[the remark this answers is not in the document]
Comment (warpfork) [reply registry.go:20]: Thank you for bringing up this whole range of questions.  (They are good.  So far, I made these diffs while trying to evade them!  But that does not represent a stance so much as it was just short-term diff-scope-minimization.)

This came up in enough places throughout this PR that I opened a whole new issue to talk through it: https://github.com/multiformats/go-multihash/issues/137

File: registry.go [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+// https://github.com/multiformats/multicodec/blob/master/table.csv .
+//
+// If Register is called with the same indicator code more than once, the last call wins.
+// In practice, this means that if an application has a strong opinion about what implementation to use for a certain hash
+// (e.g., perhaps they want to override the sha256 implementation to use a special hand-rolled assembly variant
+// rather than the stdlib one which is registered by default),
+// then this can be done by making a Register call with that effect at init time in the application's main package.
+// This should have the desired effect because the root of the import tree has its init time effect last.
+func Register(indicator uint64, hasherFactory func() hash.Hash) {
+	registry[indicator] = hasherFactory
```
[the remark this answers is not in the document]
Comment (warpfork) [reply registry.go:36]: Fair question.  I tend not to write nility checks like that, but I don't have a strongly held stance on the issue.

File: register/sha3/multihash_sha3.go (hunk 11 above)
Comment (warpfork) [reply register/sha3/multihash_sha3.go:25]: Made https://github.com/multiformats/go-multihash/issues/137 to discuss how to solve.

File: registry.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (aschmahmann) [reply registry.go:54]: Well it also returns `ErrSumNotSupported`, for example if the code is defined but the hash function isn't.

https://github.com/multiformats/go-multihash/blob/6f1ea18f1da5f7735ea31b5e2011da61c409e37f/sum.go#L51

For example, keccak-384 or x11 have codes, but no corresponding automatically registered function

https://github.com/multiformats/go-multihash/blob/6f1ea18f1da5f7735ea31b5e2011da61c409e37f/multihash.go#L135

File: errata.go (hunk 3 above)
Comment (Stebalien) [reply errata.go:33]: > I thought about (2), too, but it would make our Sum method side-effecting (ex: two consecutive calls would yield different results) which would also be a break of what Sum is expected to do.

No?

File: register/blake2/multihash_blake2.go (hunk 8 above)
Comment (Stebalien) [reply register/blake2/multihash_blake2.go:28]: Yeah, I guess... We can always fix this later if necessary.

Maybe don't export them then?

File: register/murmur3/multihash_murmur3.go (hunk 10 above)
Comment (Stebalien) [reply register/murmur3/multihash_murmur3.go:24]: Removing it sounds reasonable.

File: sum.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (aschmahmann) [reply sum.go:34]: As pointed out in https://github.com/multiformats/go-multihash/issues/137#issuecomment-791516068 at least the case of DefaultLengths seems easy since we can use `hash.Size()`. That probably means the DefaultLengths map can be dropped as it is replaceable with `h, err: = GetHasher(code) ... err things ... h.Size()`

File: errata.go (hunk 3 above)
Comment (warpfork) [reply errata.go:33]: ... Right, I figured out the correct sprinkling of `[0:0]`s.  I was clearly too tired when I looked at this last time.  It now uses the `digest` memory repeatedly.

This wasn't my original goal or focus, but I believe this is now significantly _improved_ from the previous implementation, which returned new slices in at least two places.

File: register/blake2/multihash_blake2.go (hunk 8 above)
Comment (warpfork) [reply register/blake2/multihash_blake2.go:28]: Done.

File: registry.go [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+//
+// The indicator code should be per the numbers reserved and described in
+// https://github.com/multiformats/multicodec/blob/master/table.csv .
+//
+// If Register is called with the same indicator code more than once, the last call wins.
+// In practice, this means that if an application has a strong opinion about what implementation to use for a certain hash
+// (e.g., perhaps they want to override the sha256 implementation to use a special hand-rolled assembly variant
+// rather than the stdlib one which is registered by default),
+// then this can be done by making a Register call with that effect at init time in the application's main package.
+// This should have the desired effect because the root of the import tree has its init time effect last.
```
[the remark this answers is not in the document]
Comment (warpfork) [reply registry.go:34]: Yep.

> If a package has imports, the imported packages are initialized before initializing the package itself.

- https://golang.org/ref/spec#Package_initialization

File: registry.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (warpfork) [reply registry.go:20]: ... okay, my time & energy budget for recursing up through that concern tree has expired.

- I'm going to keep `uint64` as a map key (rather than any typedefs) because that's what this repo did before;
- I'm not going to add the https://github.com/multiformats/go-multicodec repo for symbols right now because it's not in this repo right now;
- I'm going to replace all the current use of bare constants with the named constants already in this package and leave them as they are.
- I'm going to make a concerted effort to do nothing to the `Names` and `Codes` maps in the rest of this PR because the things we'd probably want to do with them would almost certainly involve bringing in the above, which... E_SCOPE_LIMIT.

Additional boat rocking about constants can come some other day and some other PR.

File: register/murmur3/multihash_murmur3.go (hunk 10 above)
Comment (warpfork) [reply register/murmur3/multihash_murmur3.go:24]: Sold.

File: sum.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (warpfork) [reply sum.go:34]: Right, yeah, using `Hash.Size` does seem to do the thing here.

I drafted the diff that removes `DefaultLengths` and it's very small and clean, so I suggest we go with dropping it.  It seems very unlikely anyone else will have been touching that map directly.

File: sum_test.go [hunk 23]
```diff
@@ -122,75 +123,22 @@ func BenchmarkBlake2B(b *testing.B) {
...
-		if err == nil {
-			t.Fatal(fmt.Sprintf("identity hash of length %d smaller than data length %d didn't fail",
-				l, dataLength))
-		}
-	}
-}
-
-// Ensure that invalid codecs can't be registered, and existing hash funcs
-// won't be overwritten.
-func TestRegisterHashFunc(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (warpfork) [reply sum_test.go:157]: I dunno, that seems pretty low value to me; it would basically be testing golang maps work as expected.  There's just not that much logic in the `Register` function anymore.

File: sum.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (warpfork) [reply sum.go]: ... I take it back, there's one more usage I didn't see:

https://github.com/multiformats/go-multihash/blob/6f1ea18f1da5f7735ea31b5e2011da61c409e37f/opts/opts.go#L113-L115

- (Sidenote) That's an ...interesting... piece of code
  - Not sure quietly setting the value to a smaller number is what I'd expect from that.  (But not necessarily here to argue about it either.)
  - Not sure "default length" and "max length" are really the same concepts.  (Meh, naming.)
- It's wanting to get this size info but not actually about to do any hashing yet, so it would rather not experience an allocation just to get the property information.

Thoughts?  Is this worth having an length info map for, or should we do something else here?

File: errata.go (hunk 1 above)
Comment (Stebalien) [reply errata.go:13]: :coffee:

Comment (Stebalien) [reply errata.go:13]: (maybe go will just optimize this one away?)

Comment (warpfork) [reply errata.go:13]: yeah if there's no variable name for the receiver at all I'm pretty sure there's no memory being shuffled.

(I made the other methods on this type take pointer receivers.)

(I kinda doubt it matters anyway, IME/ISTM that roughly three words is about where one starts seeing a difference, and that's about how big bytes.Buffer is.  But it probably depends on how you use it, I'm not really caring enough to be counting these nanos, etc, etc.)

File: sum.go (hunk 13 above)
Comment (warpfork) [reply sum.go:26]: Yeah, I was smitten with wishful thinking here -- I _thought_ had heard something about the compiler getting clever enough to sometimes see through interfaces and avoid the escapes.  I was probably thinking of [this](https://twitter.com/mdempsky/status/1321894360116420608?s=09)/[this](http://golang.org/issue/33160), though, which is of limited scope and indeed doesn't apply here.

I'll drop this to just using 'nil' again so the hash function auto-sizes it.  Same effect.

**Comment by willscott** [comment]
is this ready for a round of review? are there any loose things that still need to happen?

File: errata.go (hunk 1 above)
Comment (Stebalien) [reply errata.go:13]: We should still make these consistent (I'm pretty sure tooling will complain, but the tooling is currently broken due to the build issues).

File: multihash.go (hunk 6 above)
Comment (warpfork) [reply multihash.go:289]: I think we'd need to clarify what this function is _for_, if we reintroduce it.

If I add back a function called `KnownCode`, and the body checks the new `registry` map for a implementation existing, and call it in all the places `ValidCode` was previously called... I get test failures (for [these](https://github.com/multiformats/go-multihash/blob/6f1ea18f1da5f7735ea31b5e2011da61c409e37f/multihash_test.go#L59-L60), though I think there may be... legitimate issues with those entries).)

If the idea is to check whether the number could describe a multihash at all (rather than whether we have an implementation registered for it), it would be better to implement it based on go-multicodec, but that recurses into #137 town, [which I no longer want to wrangle in this PR](https://github.com/multiformats/go-multihash/pull/136#discussion_r588875295) because the scope creep is slowing this down.

In either case, I think this function's core concept is questionable, because it swims against "be a data carrier" and "forward compatible" as an objective.

Comment (Stebalien) [reply multihash.go:289]: I agree. I think "supported" may be useful (i.e., can we compute a hash with this code), but we can add that later.

File: sum_test.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (Stebalien) [reply sum_test.go:128]: It ensures that ID hashes _cannot_ truncate (nonsensical). We passed the hash length to the hasher specifically to allow this check.

I guess we have two easy solutions:

1. Make the hasher constructor take a size. We'd probably need two register functions to handle this (`Register` and `RegisterSized`, or something like that).
2. Add an optional `MinSize() int` function to our hashers (i.e., type assert to see if this is defined). When truncating, check to see if there is a `MinSize()` defined and, if there is, make sure the requested size is at least the minimum size.
3. Allow _registering_ a minimum size for a codec.

2 or 3 probably make the most sense where 3 is likely the most performant.

File: registry.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (Stebalien) [reply registry.go:54]: Well, "not supported" is correct because we don't support that hash function. But yeah, we should indicate what isn't supported. I believe we did it this way before because go didn't have a way to check any form of `errors.Is`. Now that we have that, we can define some base error and derive a new error from that.

[the remark this answers is not in the document]
Comment (warpfork) [reply registry.go]: Alright, I've done this, wrapping-style, in cbd218cd90910541ab7a67d6a52ba999716ff2d7 .

I tried to keep it as low impact as possible (e.g. keep an `ErrSumNotSupported` value, by that symbol name).  It works, but I wanna jot some notes that I have tepid feelings about how it turns out: when using wrap like this to provide more info, consumers *still* have to change if they really cared about the previous exact error value, because now they have to use `errors.Is` (as demonstrated by the diff to one of our own tests).

If I'd ditched the goal to keep an `ErrSumNotSupported` value, I probably would've written this by introducing a new struct type to contain the code detail... but in terms of downstream effects, it would be sort of potæto/potato: the exact syntax that consumers would need to decide the "type" of error would be different, but it would still be a diff.

I guess there's really no way to do changes to errors in golang that aren't going to break something (and, we don't really get compiler help detecting it, either, because yay, golang), and this is just life.

In this case I'm hoping it just won't be a big problem in practice, for all the reasons already covered up-thread.

File: sum_test.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (warpfork) [reply sum_test.go:128]: Okay, I'm going to revive this test and push a change to the `Sum` function which explicitly rejects truncations for identity multihashes.  That's the most conservative change that gets this test back and gets it passing and maintains the old behavior and its limits.

The MinSize idea sounds interesting (especially in light of https://github.com/ipfs/go-verifcid/ , which iirc is also concerned about overly short hashes) but sounds like it'll broach a bigger discussion than I wanna hitch to this PR.

**Comment by warpfork** [comment]
Okay!  This is finally green.

I believe all comments are addressed.

Last call for any more reviews?

File: registry.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (warpfork) [reply registry.go:40]: Decided to add this check.  It's cheap.  Why not.

**Comment by warpfork** [comment]
I've attempted to push this downstream through go-cid, and here's what that might look like: https://github.com/ipfs/go-cid/commit/26b387af822073023f9c026a4ea68a6349abf20e#diff-88c2c4b0892fb0172322544e3937e937bb18beccecb1b24a1b5028758a43685bR346-R375

There's some questions worth a quick review there.  Namely: it's entirely possible to make all the code work without a need for `DefaultLengths`, but there are some tests which conflict with it.  It's not clear to me if those tests are testing something important and relied upon or happen to assert what they do incidentally.

Everything else looks fine -- although we do again end up seeing the `go-multihash/register/all` import now needing to appear in test files, which should probably not be a surprise.

**Comment by Stebalien** [comment]
> There's some questions worth a quick review there. Namely: it's entirely possible to make all the code work without a need for DefaultLengths, but there are some tests which conflict with it. It's not clear to me if those tests are testing something important and relied upon or if they happen to assert what they do incidentally.

We _do_ need DefaultLengths, we use it in quite a few places. You can find these places by:

1. Updating go-multihash in go-ipfs, lotus, etc.
2. Seeing what breaks.

Let's make this PR _minimally breaking_ so we can merge it ASAP and move on.

**Comment by warpfork** [comment]
`DefaultLengths` is back in.

Here's what propagating through go-cid looks like now, with all tests passing: https://github.com/ipfs/go-cid/pull/118

**Comment by warpfork** [comment]
I guess the final check here is: do we want to keep `github.com/multiformats/go-multihash`, the root package of this repo, still dragging in all transitive dependencies, to really really minimize change?  E.g. effectively have it import `register/all` and `register/miniosha256`?

https://github.com/multiformats/go-multihash/issues/131 did originally describe doing the registration rework in a new package, and keeping the go-multihash root unchanged.

As it stands now, the go-multihash *symbols* exported are pretty much unchanged, but the code does go all-in on making the dependencies optional, so the behavior does shift.  If we're scared of this and the consequent need to add the registration imports in downstream applications, and we'd rather make the dependency minimization an opt-in action that has to be done separately from just updating the library by putting it in a deeper package, we can do that.

... heck, I'll just draft that now.  If we introduce a new package, who wants to name it?  `go-multihash/core`?  or `go-multihash/lite`?  or just put it in `go-mulithash/register`?

**Comment by Stebalien** [comment]
Yeah, let's minimize breakage This library is a bit popular: https://pkg.go.dev/github.com/multiformats/go-multihash?tab=importedby. A "core" package would work (I'd call it that).

I guess the question is: what do _you_ need. This appears to be blocking other work, so I want to make sure we merge _that_ need ASAP.

If you don't really need to reduce dependencies (I'm pretty sure you don't, at least right now) then I'd:

1. Import these deps from the root for now.
2. Release this as a 1.0 release.
3. Immediately start on a 2.0 release that _doesn't_ "register" everything.
4. Make the 1.0 release depend on the 2.0 release.

Basically, we need unblock this ASAP and we need to avoid causing a meltdown. Let's just go with the easiest approach and introduce a 2.0 (we'll get there eventually regardless).

**Comment by warpfork** [comment]
Alright.  `go-multihash/core` is now a package, it does the thing.  `go-mulithash` still brings on all the deps unconditionally.

I think I've run out of changes to avoid.  `go-cid` can now be updated to this last commit with literally no code changes.

I'll also bow out on version tagging.  Whatever numbers you like, fine by me.  (I'm wary of major numbers and the Different Behavior they prompt from `go mod`, but... what could go wrong, really)

**Comment by Stebalien** [comment]
I've tested with go-ipfs and the only broken test is the one that tries to use murmur (to make sure we don't allow it).

**Review by Stebalien — APPROVED** [review]

**Comment by Stebalien** [comment]
@warpfork could you cut a minor release when ready?

**Comment by warpfork** [comment]
This is now available within the `v0.0.15` tag.
